src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs

```diff
@@ -5,23 +5,152 @@
 
 namespace System.Text.RegularExpressions
 {
+    /// <summary>
+    /// An exception as a result of a parse error in a regular expression <see cref="RegularExpressions"/>, with
+    /// detailed information in the <see cref="Error"/> and <see cref="Offset"/> properties.
+    /// </summary>
     [Serializable]
-    internal sealed class RegexParseException : ArgumentException
+    public sealed class RegexParseException : ArgumentException
     {
-        private readonly RegexParseError _error; // tests access this via private reflection
-
         /// <summary>Gets the error that happened during parsing.</summary>
-        public RegexParseError Error => _error;
+        public RegexParseError Error { get; }
 
-        /// <summary>Gets the offset in the supplied pattern.</summary>
+        /// <summary>Gets the zero-based character offset in the regular expression pattern where the parse error occurs.</summary>
         public int Offset { get; }
 
-        public RegexParseException(RegexParseError error, int offset, string message) : base(message)
+        internal RegexParseException(RegexParseError error, int offset, string message) : base(message)
+        {
+            Error = error;
+            Offset = offset;
+        }
+
+        internal RegexParseException(string pattern, RegexParseError error, int offset) : base(MakeMessage(pattern, error, offset))
         {
-            _error = error;
+            Error = error;
             Offset = offset;
         }
 
+        /// <summary>
+        /// Construct a <see cref="RegexParseException"/> that creates a default message based on the given <see cref="RegexParseError"/> value.
+        /// </summary>
+        /// <param name="pattern">The pattern of the regular expression.</param>
+        /// <param name="error">The <see cref="RegexParseError"/> value detailing the type of parse error.</param>
+        /// <param name="offset">The zero-based offset in the regular expression where the parse error occurs.</param>
+        private static string MakeMessage(string pattern, RegexParseError error, int offset)
```

Comment:
Don't change this PR for this! 😄  but for future changes the switch pattern syntax can be more compact? I think it's something like

```c#
    string resource = error switch
    {
        RegexParseError.Unknown => SR.Format(SR.MakeException, pattern, offset, SR.Generic),
        RegexParseError.AlternationHasTooManyConditions => SR.Format(SR.MakeException, pattern, offset, SR.AlternationHasTooManyConditions),
        //...
    };
```

Reply:
Tx. Yeah, I considered it, just call me old-fashioned (and the default helper screws up the code big-time). No worries, I won't commit a fix ;)